### DISCUSSION TITLE
Certificate not being loaded in Azure App Service

### DISCUSSION THREAD

**Discussion by rmcelroy-hylaine** [pnp/pnpcore#1151] [Q&A] [answered]
Hi all - I hate coming to the community board with such a simple question. 

I've developed a .NET 6.0 REST API hosted on an Azure App Service. When hitting endpoints that use PnP Core, I'm getting the error:

`"Value cannot be null. (Parameter 'certificate')"`

I can replicate this error locally when the certificate is not loaded or the thumbprint is malformed. The app service has the certificate uploaded, and appsettings has been updated to load all certificates:
![image](https://user-images.githubusercontent.com/106849139/229133008-a54e3953-a3d1-4911-93b3-36390d79b2a2.png)

![image](https://user-images.githubusercontent.com/106849139/229133122-13d9512f-7fe1-4344-a180-653f080a1b84.png)

Code to load the cert is straighforward. I'm wondering if we need to change StoreLocation based on if we're running locally or deployed in the cloud?

`

            services.AddPnPCore();
            services.Configure<PnPCoreOptions>(Configuration.GetSection("PnPCore"));            
            services.AddPnPCoreAuthentication(
                options =>
                {
                    options.Credentials.Configurations.Add("x509Certificate", new PnPCoreAuthenticationCredentialConfigurationOptions
                    {
                        ClientId = Configuration.GetValue<string>("Sharepoint:ApplicationId"),
                        TenantId = Configuration.GetValue<string>("Sharepoint:TenantId"),
                        X509Certificate = new PnPCoreAuthenticationX509CertificateOptions
                        {
                            StoreLocation = StoreLocation.LocalMachine,
                            StoreName = StoreName.My,
                            Thumbprint = Configuration.GetValue<string>("Sharepoint:CertThumbprint")
                        }
                    });

                    options.Credentials.DefaultConfiguration = "x509Certificate";
                    options.Sites.Add("SiteToWorkWith", new PnPCoreAuthenticationSiteOptions
                    {
                        AuthenticationProviderName = "x509Certificate"
                    });

                }
            );
`

App service itself indicates no problems with the cert:
![image](https://user-images.githubusercontent.com/106849139/229134119-06f22703-c8d4-49fc-967f-59a452b818ea.png)

Previous research in other PnP related github areas include:

* https://github.com/SharePoint/PnP-Partner-Pack/issues/184
* https://github.com/pnp/PnP/issues/1401

Anything I'm missing here? Once again, I apologize - just not sure how to check to see how this cert is being loaded in the app service.

**Comment by jansenbe** [comment]
Can you check this tutorial, it explains the scenario you need: https://pnp.github.io/pnpcore/tutorials/azurefunctions/v4processisolatedapponly.html

**Reply by rmcelroy-hylaine** [reply]
Ugh, yes this is helpful. That example presumes obtaining the cert from a KeyVault, which is what we'll do longer term.

When the cert is stored on the host, though, it looks like using the `X509Store(StoreName.My, StoreLocation.CurrentUser)` should work. Looks like we're using LocalMachine which is what I'll need to change based on the configuration.

https://learn.microsoft.com/en-us/azure/app-service/configure-ssl-certificate-in-code#load-certificate-in-windows-apps

Thanks for the follow up!

**Comment by rmcelroy-hylaine** [comment, marked as answer]
I wanted to go ahead and close this issue. The problem indeed was the `StoreLocation` parameter.

Locally, you want this to be `StoreLocation.LocalMachine` if you install the cert through the wizard with default parameters. In the Cloud, it must be `StoreLocation.CurrentUser`. Final Code:

`

	public void ConfigureServices(IServiceCollection services)
	{
		var isDevelopment = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == "Development";
		//...
		services.AddPnPCore();
		services.Configure<PnPCoreOptions>(Configuration.GetSection("PnPCore"));

		var certificateStoreLocation = isDevelopment ? StoreLocation.LocalMachine : StoreLocation.CurrentUser;
		services.AddPnPCoreAuthentication(
		    options =>
		    {
		        options.Credentials.Configurations.Add("x509Certificate", new PnPCoreAuthenticationCredentialConfigurationOptions
		        {
		            ClientId = Configuration.GetValue<string>("Sharepoint:ApplicationId"),
		            TenantId = Configuration.GetValue<string>("Sharepoint:TenantId"),
		            X509Certificate = new PnPCoreAuthenticationX509CertificateOptions
		            {
		                StoreLocation = certificateStoreLocation,
		                StoreName = StoreName.My,
		                Thumbprint = Configuration.GetValue<string>("Sharepoint:CertThumbprint")
		            }
		        });

		        options.Credentials.DefaultConfiguration = "x509Certificate";
		        options.Sites.Add("SiteToWorkWith", new PnPCoreAuthenticationSiteOptions
		        {
		            AuthenticationProviderName = "x509Certificate"
		        });

		    }
		);
	}
`
